test/distributed/_tensor/test_attention.py

```diff
@@ -40,137 +38,146 @@
 class RingAttentionTest(DTensorTestBase):
     @property
     def world_size(self) -> int:
-        return 2
+        return torch.cuda.device_count()
 
     @skip_if_lt_x_gpu(2)
     @skipIfRocm  # Missing _c10d_functional_autograd::all_to_all_single
     @unittest.skipIf(
-        not PLATFORM_SUPPORTS_FLASH_ATTENTION, "Does not support flash attention"
+        not PLATFORM_SUPPORTS_FUSED_ATTENTION,
+        "Does not support flash nor efficient attention",
     )
     @with_comms
     @parametrize("is_causal", [True, False])
-    def test_ring_attention_sdpa(self, is_causal: bool) -> None:
-        device_mesh = DeviceMesh(
-            self.device_type,
-            torch.arange(0, self.world_size),
-        )
+    @parametrize("compiled", [True, False])
+    @parametrize("backend", backends)
+    def test_ring_attention_sdpa(
+        self, is_causal: bool, compiled: bool, backend: SDPBackend
+    ) -> None:
+        device_mesh = DeviceMesh(self.device_type, torch.arange(0, self.world_size))
         dtype = torch.bfloat16
         bs = 8
-        query_tokens = 8
-        context_tokens = query_tokens if is_causal else 8
+        query_tokens = 64
+        context_tokens = 64
         dim = 32
         nheads = 8
-        query = torch.rand(
+        torch.manual_seed(10)
+        dtype = (
+            torch.bfloat16 if backend == SDPBackend.FLASH_ATTENTION else torch.float32
+        )
+
+        if is_causal and compiled and self.world_size > 2:
+            # TODO: Fix this after we move `wait_tensor` to use `with_effect`.
+            return
+
+        q = torch.rand(
             (bs, nheads, self.world_size * query_tokens, dim),
             device=self.device_type,
             dtype=dtype,
             requires_grad=True,
         )
-        key = torch.rand(
+        k = torch.rand(
             (bs, nheads, self.world_size * context_tokens, dim),
             device=self.device_type,
             dtype=dtype,
+            requires_grad=True,
         )
-        value = torch.rand(
+        v = torch.rand(
             (bs, nheads, self.world_size * context_tokens, dim),
             device=self.device_type,
             dtype=dtype,
+            requires_grad=True,
         )
 
-        query_placement = [Shard(2)]
-        dquery = distribute_tensor(query, device_mesh, query_placement)
-        self.assertEqual(query.shape, (bs, nheads, self.world_size * query_tokens, dim))
-
-        context_placement = [Shard(2)]
-        dkey = distribute_tensor(key, device_mesh, context_placement)
-        dvalue = distribute_tensor(value, device_mesh, context_placement)
-        for t in [dkey, dvalue]:
-            self.assertEqual(
-                t.shape, (bs, nheads, context_tokens * self.world_size, dim)
-            )
-            self.assertEqual(t.to_local().shape, (bs, nheads, context_tokens, dim))
+        # Ensure all ranks have the same initialization data.
+        with torch.no_grad():
+            dist.broadcast(q, src=0)
```

Comment:
iiuc, here we are switching to testing the ring attention without sharding the inputs, to verify the local sdpa is numerically the same with local ring attention? 🤔 but the ring attention works in the way that it's calculating the global attention score that shards qkv on sequence dimension already, so is this comparison something we want to do?

I thought the old way of creating global qkv, shard them on sequence_dim, then comparing global spda results with ring attention should be the correct numeric test setup?

Reply:
There should be no differences between these two tests. The way current test is to get the global SDPA result and rank0 gets the first half and rank1 gets the second half. This should match the local SDPA result using because we shard on the sequence dimension.

But ye, it may be more straightforward to just compare the global results but that requires the reshuffling which is introduced in the next PR. The test will be changed to compare the global SDPA result in the next PR. I don't want to merge everything into one PR, so bear with this test for just a while.